src/later.cpp

```diff
@@ -27,21 +23,30 @@ class ProtectCallbacks {
   }
 };
 
-// Save a call expression as NFramesCallback. This is called at startup.
-// [[Rcpp::export]]
-void saveNframesCallback(SEXP exp) {
-  // TODO: Is R_PreserveObject necessary here?
-  R_PreserveObject(exp);
-  
-  nframes = exp;
+// Returns number of frames on the call stack. Basically just a wrapper for
+// base::sys.nframe().
+int sys_nframe() {
+  SEXP e, result;
+  int errorOccurred;
+
+  PROTECT(e = Rf_lang1(Rf_install("sys.nframe")));
+  result = R_tryEval(e, R_BaseEnv, &errorOccurred);
+
+  if (errorOccurred) {
+    UNPROTECT(1);
```

Comment:
When doing C-style manual resource management (i.e. without destructors) I think it's customary to try to have a single point of return for the function, so the cleanup code is in one place.